# Final Project

Group 4: Jiaqi Bai, Somender Chaudhary, Shreyas Deo, Andrew Levy, Wenqwing Qin, Matias Roca-Guiulfo, and Alexis Yang

## Ask 1 - Data Collection (10%)

Use the JHU datasets website, collect the following dataset 
- Cases and Deaths (04/12/2020 – 11/15/2021)
- Vaccine and People Vaccinated (12/10/2020 – 11/15/2021)
- Testing

The readme/dictionary files are available for all datasets Cases and Deaths, Vaccine and Testing.

Acquire daily U.S. report. Perform initial exploration to make sure it is suitable for dimensional modeling and analytical analysis.

Expectation: Provide a list of your data sources, The links to the data sources and links to any documentations you used.

Links
- JHU Datasets: https://coronavirus.jhu.edu/about/how-to-use-our-data
- Cases: https://github.com/CSSEGISandData/COVID-19/tree/master/csse_covid_19_data/csse_covid_19_daily_reports_us
- Deaths: https://github.com/govex/COVID-19/blob/master/data_tables/vaccine_data/us_data/time_series/vaccine_data_us_timeline.csv & https://raw.githubusercontent.com/govex/COVID-
- Vaccine: https://github.com/govex/COVID-19/blob/master/data_tables/vaccine_data/us_data/time_series/vaccine_data_us_timeline.csv
- People Vaccinated: https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/vaccine_data/us_data/time_series/people_vaccinated_us_timeline.csv
- Testing: https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/testing_data/time_series_covid19_US.csv
Readme/dictionary files
- https://github.com/CSSEGISandData/COVID-19/blob/master/csse_covid_19_data/README.md
- https://github.com/govex/COVID-19/blob/master/data_tables/vaccine_data/us_data/data_dictionary.csv
- https://github.com/govex/COVID-19/blob/master/data_tables/testing_data/data_dictionary.csv

#### Collecting Cases and Deaths Dataset 

#### Collecting Cases and Deaths Dataset 

In [92]:
!wc -l cases_deaths.csv # printing the number of lines presents in the cases_deaths.csv file

34471 cases_deaths.csv


In [93]:
!head cases_deaths.csv # printing the first 10 lines of the cases_deaths.csv file

Province_State,Country_Region,Last_Update,Lat,Long_,Confirmed,Deaths,Recovered,Active,FIPS,Incident_Rate,Total_Test_Results,People_Hospitalized,Case_Fatality_Ratio,UID,ISO3,Testing_Rate,Hospitalization_Rate
Alabama,US,2021-01-02 05:30:44,32.3182,-86.9023,365747,4872,202137.0,158738.0,1.0,7459.375895463867,3275341.0,,1.3320683423240656,84000001.0,USA,66800.27369964625,
Alaska,US,2021-01-02 05:30:44,61.3707,-152.4044,47019,206,7165.0,39615.0,2.0,6422.844801071706,1275750.0,,0.4384284680543141,84000002.0,USA,174391.1857780451,
American Samoa,US,2021-01-02 05:30:44,-14.271,-170.132,0,0,,0.0,60.0,0.0,2140.0,,,16.0,ASM,3846.084721698028,
Arizona,US,2021-01-02 05:30:44,33.7298,-111.4312,530267,9015,76934.0,444318.0,4.0,7285.171274003372,5155330.0,,1.7000869373353424,84000004.0,USA,39551.860582022906,
Arkansas,US,2021-01-02 05:30:44,34.9697,-92.3731,229442,3711,199247.0,26484.0,5.0,7602.945718144717,2051488.0,,1.6174022192972517,84000005.0,USA,67979.49767446793,
California,US,2021-01-02 05:30:

In [94]:
!head -5 cases_deaths.csv | csvlook # rendering the cases_deaths.csv file into a readable format 

| Province_State | Country_Region |         Last_Update |      Lat |     Long_ | Confirmed | Deaths | Recovered |  Active | FIPS | Incident_Rate | Total_Test_Results | People_Hospitalized | Case_Fatality_Ratio |        UID | ISO3 | Testing_Rate | Hospitalization_Rate |
| -------------- | -------------- | ------------------- | -------- | --------- | --------- | ------ | --------- | ------- | ---- | ------------- | ------------------ | ------------------- | ------------------- | ---------- | ---- | ------------ | -------------------- |
| Alabama        | US             | 2021-01-02 05:30:44 |  32.318… |  -86.902… |   365,747 |  4,872 |   202,137 | 158,738 |    1 |    7,459.376… |          3,275,341 |                     |              1.332… | 84,000,001 | USA  |  66,800.274… |                      |
| Alaska         | US             | 2021-01-02 05:30:44 |  61.371… | -152.404… |    47,019 |    206 |     7,165 |  39,615 |    2 |    6,422.845… |          1,275,750 |                     | 

In [149]:
!head -n 1000 cases_deaths.csv | csvstat # printing the descriptive statistics for the first 1000 lines of the cases_deaths.csv file 

  1. "Province_State"

	Type of data:          Text
	Contains null values:  False
	Unique values:         58
	Longest value:         24 characters
	Most common values:    Alabama (18x)
	                       Alaska (18x)
	                       American Samoa (18x)
	                       Arizona (18x)
	                       Arkansas (18x)

  2. "Country_Region"

	Type of data:          Text
	Contains null values:  False
	Unique values:         1
	Longest value:         2 characters
	Most common values:    US (999x)

  3. "Last_Update"

	Type of data:          DateTime
	Contains null values:  False
	Unique values:         29
	Smallest value:        2021-01-02 05:30:44
	Largest value:         2021-01-19 09:30:34
	Most common values:    2021-01-02 05:30:44 (58x)
	                       2021-01-03 05:30:33 (58x)
	                       2021-01-04 05:30:45 (58x)
	                       2021-01-05 05:30:33 (58x)
	                       2021-01-06 05:30:39 (58x)

  4. "Lat"

	Type of data:

#### Collecting Vaccine Dataset

In [6]:
!wget https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/vaccine_data/us_data/time_series/vaccine_data_us_timeline.csv

--2021-11-27 13:59:18--  https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/vaccine_data/us_data/time_series/vaccine_data_us_timeline.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.111.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10315455 (9.8M) [text/plain]
Saving to: 'vaccine_data_us_timeline.csv’

vaccine_data_us_tim 100%[===================>]   9.84M   362KB/s    in 27s     

2021-11-27 13:59:47 (366 KB/s) - 'vaccine_data_us_timeline.csv’ saved [10315455/10315455]



In [7]:
!mv vaccine_data_us_timeline.csv vaccine.csv

In [2]:
!wc -l vaccine.csv

115883 vaccine.csv


In [3]:
!head vaccine.csv | csvcut -n

  1: Province_State
  2: Date
  3: Vaccine_Type
  4: FIPS
  5: Country_Region
  6: Lat
  7: Long_
  8: Doses_alloc
  9: Doses_shipped
 10: Doses_admin
 11: Stage_One_Doses
 12: Stage_Two_Doses
 13: Combined_Key


In [4]:
!head -5 vaccine.csv | csvlook

| Province_State |       Date | Vaccine_Type | FIPS | Country_Region |     Lat |    Long_ | Doses_alloc | Doses_shipped | Doses_admin | Stage_One_Doses | Stage_Two_Doses | Combined_Key |
| -------------- | ---------- | ------------ | ---- | -------------- | ------- | -------- | ----------- | ------------- | ----------- | --------------- | --------------- | ------------ |
| Alabama        | 2020-12-10 | All          | True | US             | 32.318… | -86.902… |             |               |             |                 |                 | Alabama, US  |
| Alabama        | 2020-12-10 | Moderna      | True | US             | 32.318… | -86.902… |             |               |             |                 |                 | Alabama, US  |
| Alabama        | 2020-12-10 | Pfizer       | True | US             | 32.318… | -86.902… |             |               |             |                 |                 | Alabama, US  |
| Alabama        | 2020-12-11 | All          | True | US         

In [5]:
!head -n 1000 vaccine.csv | csvstat

  1. "Province_State"

	Type of data:          Text
	Contains null values:  False
	Unique values:         1
	Longest value:         7 characters
	Most common values:    Alabama (999x)

  2. "Date"

	Type of data:          Date
	Contains null values:  False
	Unique values:         198
	Smallest value:        2020-12-10
	Largest value:         2021-06-25
	Most common values:    2021-03-15 (6x)
	                       2021-03-16 (6x)
	                       2021-03-17 (6x)
	                       2021-03-18 (6x)
	                       2021-03-19 (6x)

  3. "Vaccine_Type"

	Type of data:          Text
	Contains null values:  False
	Unique values:         6
	Longest value:         10 characters
	Most common values:    All (198x)
	                       Moderna (198x)
	                       Pfizer (198x)
	                       Unassigned (194x)
	                       Janssen (109x)

  4. "FIPS"

	Type of data:          Boolean
	Contains null values:  False
	Unique values:         1
	Most

#### Collecting People Vaccinated Dataset

In [12]:
!wget https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/vaccine_data/us_data/time_series/people_vaccinated_us_timeline.csv

--2021-11-27 14:01:27--  https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/vaccine_data/us_data/time_series/people_vaccinated_us_timeline.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.111.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1581333 (1.5M) [text/plain]
Saving to: 'people_vaccinated_us_timeline.csv’

people_vaccinated_u 100%[===================>]   1.51M   974KB/s    in 1.6s    

2021-11-27 14:01:30 (974 KB/s) - 'people_vaccinated_us_timeline.csv’ saved [1581333/1581333]



In [13]:
!mv people_vaccinated_us_timeline.csv people_vaccinated.csv

In [14]:
!wc -l people_vaccinated.csv

21373 people_vaccinated.csv


In [59]:
!head people_vaccinated.csv | csvcut -n

  1: FIPS
  2: Province_State
  3: Country_Region
  4: Date
  5: Lat
  6: Long_
  7: Combined_Key
  8: People_Fully_Vaccinated
  9: People_Partially_Vaccinated


In [16]:
!head -5 people_vaccinated.csv | csvlook

/bin/bash: csvlook: command not found
head: write error: Broken pipe


In [61]:
!head -n 1000 people_vaccinated.csv | csvstat

  1. "FIPS"

	Type of data:          Number
	Contains null values:  False
	Unique values:         3
	Smallest value:        1
	Largest value:         60
	Sum:                   18,873
	Mean:                  18.892
	Median:                2
	StDev:                 26.755
	Most common values:    1 (351x)
	                       2 (351x)
	                       60 (297x)

  2. "Province_State"

	Type of data:          Text
	Contains null values:  False
	Unique values:         3
	Longest value:         14 characters
	Most common values:    Alabama (351x)
	                       Alaska (351x)
	                       American Samoa (297x)

  3. "Country_Region"

	Type of data:          Text
	Contains null values:  False
	Unique values:         1
	Longest value:         2 characters
	Most common values:    US (999x)

  4. "Date"

	Type of data:          Date
	Contains null values:  False
	Unique values:         351
	Smallest value:        2020-12-10
	Largest value:         2021-11-25
	Most c

#### Collecting Testing Dataset 

In [19]:
!wget https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/testing_data/time_series_covid19_US.csv

--2021-11-27 14:05:12--  https://raw.githubusercontent.com/govex/COVID-19/master/data_tables/testing_data/time_series_covid19_US.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.110.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2267428 (2.2M) [text/plain]
Saving to: 'time_series_covid19_US.csv’

time_series_covid19 100%[===================>]   2.16M   571KB/s    in 4.3s    

2021-11-27 14:05:18 (516 KB/s) - 'time_series_covid19_US.csv’ saved [2267428/2267428]



In [20]:
!mv time_series_covid19_US.csv testing.csv

In [6]:
!wc -l testing.csv

35509 testing.csv


In [7]:
!head testing.csv | csvcut -n

  1: date
  2: state
  3: cases_conf_probable
  4: cases_confirmed
  5: cases_probable
  6: tests_viral_positive
  7: tests_viral_negative
  8: tests_viral_total
  9: tests_antigen_positive
 10: tests_antigen_total
 11: people_viral_positive
 12: people_viral_total
 13: people_antigen_positive
 14: people_antigen_total
 15: encounters_viral_total
 16: tests_combined_total


In [8]:
!head -5 testing.csv | csvlook

|       date | state | cases_conf_probable | cases_confirmed | cases_probable | tests_viral_positive | tests_viral_negative | tests_viral_total | tests_antigen_positive | tests_antigen_total | people_viral_positive | people_viral_total | people_antigen_positive | people_antigen_total | encounters_viral_total | tests_combined_total |
| ---------- | ----- | ------------------- | --------------- | -------------- | -------------------- | -------------------- | ----------------- | ---------------------- | ------------------- | --------------------- | ------------------ | ----------------------- | -------------------- | ---------------------- | -------------------- |
| 2020-03-06 | AK    |                     |                 |                |                      |                      |                 8 |                        |                     |                       |                    |                         |                      |                        |                   

In [23]:
!head -n 1000 testing.csv | csvstat

  1. "date"

	Type of data:          Date
	Contains null values:  False
	Unique values:         600
	Smallest value:        2020-03-06
	Largest value:         2021-10-26
	Most common values:    2020-03-07 (2x)
	                       2020-03-08 (2x)
	                       2020-03-09 (2x)
	                       2020-03-10 (2x)
	                       2020-03-11 (2x)

  2. "state"

	Type of data:          Text
	Contains null values:  False
	Unique values:         2
	Longest value:         2 characters
	Most common values:    AK (600x)
	                       AL (399x)

  3. "cases_conf_probable"

	Type of data:          Number
	Contains null values:  True (excluded from calculations)
	Unique values:         872
	Smallest value:        0
	Largest value:         518234
	Sum:                   103854382
	Mean:                  105115.771255
	Median:                60399
	StDev:                 140925.470189
	Most common values:    None (11x)
	                       0 (6x)
	               

## Ask 2 - Surveying and Wrangling Data (15%)

Acquire the data, survey it and wrangle it to suitable format. You may use one or more of the following methods for acquiring and wrangling the data: 
- Unix command line tools (e.g. csvkit)
- Relational databases 
- Spark
- You may also use wrangling tool (e.g. Trifacta)

Describe any concerns with the data and changes you expect to overcome. 

Expectations: document your findings in the code. Make sure you describe each step in a complete and concise manner. 

### surveying and wrangling dataset cases_deaths.csv

#### setup database 

In [95]:
#
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [97]:
!dropdb -U student final_project

dropdb: error: database removal failed: ERROR:  database "final_project" is being accessed by other users
DETAIL:  There is 1 other session using the database.


In [153]:
!createdb -U student final_project

createdb: error: database creation failed: ERROR:  database "final_project" already exists


In [98]:
%sql postgresql://student@/final_project

In [112]:
%%sql
DROP TABLE IF EXISTS cases_deaths;

CREATE TABLE cases_deaths (
Province_State text not null,
Country_Region text not null,
Last_Update date,
Lat numeric,
Long_ numeric,
Confirmed numeric,    
Deaths numeric,
Recovered numeric,
Active numeric,
FIPS numeric,
Incident_Rate numeric,
Total_Test_Results numeric,
People_Hospitalized numeric,
Case_Fatality_Ratio numeric,
UID numeric,
ISO3 text,
Testing_Rate numeric,
Hospitalization_Rate numeric
)

 * postgresql://postgres@/final_project
Done.
Done.


[]

In [113]:
!pwd

/home/ubuntu


In [114]:
%%sql
COPY cases_deaths FROM '/home/ubuntu/cases_deaths.csv'
CSV
HEADER;
#here my location is /home/ubuntu, please change it accordingly as well.

 * postgresql://postgres@/final_project
34470 rows affected.
(psycopg2.errors.SyntaxError) syntax error at or near "#"
LINE 1: #here my location is /home/ubuntu, please change it accordin...
        ^

[SQL: #here my location is /home/ubuntu, please change it accordingly as well.]
(Background on this error at: http://sqlalche.me/e/f405)


In [115]:
%%sql
SELECT COUNT(*) FROM cases_deaths; -- query total number records of cases_deaths

 * postgresql://postgres@/final_project
1 rows affected.


[(34470,)]

In [116]:
%%sql
SELECT Province_State,Country_Region,Last_Update,Lat,Long_,Confirmed,Deaths,Recovered,Active,FIPS,Incident_Rate,Total_Test_Results,People_Hospitalized,Case_Fatality_Ratio,UID,ISO3,Testing_Rate,Hospitalization_Rate,
COUNT(*) FROM cases_deaths
group by Province_State,Country_Region,Last_Update,Lat,Long_,Confirmed,Deaths,Recovered,Active,FIPS,Incident_Rate,Total_Test_Results,People_Hospitalized,Case_Fatality_Ratio,UID,ISO3,Testing_Rate,Hospitalization_Rate
having count(*)>1; -- check if there are duplicated records in cases_deaths

 * postgresql://postgres@/final_project
1 rows affected.


[('American Samoa', 'US', None, Decimal('-14.270999999999999'), Decimal('-170.132'), Decimal('0'), Decimal('0'), None, None, Decimal('60'), Decimal('0.0'), Decimal('3'), None, None, Decimal('16'), 'ASM', Decimal('5.3917075537822825'), None, 7)]

In [117]:
%%sql
select * from cases_deaths
where province_state='American Samoa' and country_region='US' and last_update is null  and lat='-14.270999999999999' and long_='-170.132' and confirmed=0 and
deaths=0 and fips=60 and incident_rate=0.0 and total_test_results=3 and uid=16 and iso3='ASM' and testing_rate='5.3917075537822825';

 * postgresql://postgres@/final_project
7 rows affected.


[('American Samoa', 'US', None, Decimal('-14.270999999999999'), Decimal('-170.132'), Decimal('0'), Decimal('0'), None, None, Decimal('60'), Decimal('0.0'), Decimal('3'), None, None, Decimal('16'), 'ASM', Decimal('5.3917075537822825'), None),
 ('American Samoa', 'US', None, Decimal('-14.270999999999999'), Decimal('-170.132'), Decimal('0'), Decimal('0'), None, None, Decimal('60'), Decimal('0.0'), Decimal('3'), None, None, Decimal('16'), 'ASM', Decimal('5.3917075537822825'), None),
 ('American Samoa', 'US', None, Decimal('-14.270999999999999'), Decimal('-170.132'), Decimal('0'), Decimal('0'), None, None, Decimal('60'), Decimal('0.0'), Decimal('3'), None, None, Decimal('16'), 'ASM', Decimal('5.3917075537822825'), None),
 ('American Samoa', 'US', None, Decimal('-14.270999999999999'), Decimal('-170.132'), Decimal('0'), Decimal('0'), None, None, Decimal('60'), Decimal('0.0'), Decimal('3'), None, None, Decimal('16'), 'ASM', Decimal('5.3917075537822825'), None),
 ('American Samoa', 'US', None, Decimal('-14.270999999999999'), Decimal('-170.132'), Decimal('0'), Decimal('0'), None, None, Decimal('60'), Decimal('0.0'), Decimal('3'), None, None, Decimal('16'), 'ASM', Decimal('5.3917075537822825'), None),
 ('American Samoa', 'US', None, Decimal('-14.270999999999999'), Decimal('-170.132'), Decimal('0'), Decimal('0'), None, None, Decimal('60'), Decimal('0.0'), Decimal('3'), None, None, Decimal('16'), 'ASM', Decimal('5.3917075537822825'), None),
 ('American Samoa', 'US', None, Decimal('-14.270999999999999'), Decimal('-170.132'), Decimal('0'), Decimal('0'), None, None, Decimal('60'), Decimal('0.0'), Decimal('3'), None, None, Decimal('16'), 'ASM', Decimal('5.3917075537822825'), None)]

In [118]:
%%sql
delete from cases_deaths
where province_state='American Samoa' and country_region='US' and last_update is null  and lat='-14.270999999999999' and long_='-170.132' and confirmed=0 and
deaths=0 and fips=60 and incident_rate=0.0 and total_test_results=3 and uid=16 and iso3='ASM' and testing_rate='5.3917075537822825'; -- delete all these duplicated rows

 * postgresql://postgres@/final_project
7 rows affected.


[]

In [119]:
%%sql
select * from cases_deaths
where Province_State is null or country_region is null or  confirmed is null or deaths is null

 * postgresql://postgres@/final_project
0 rows affected.


[]

### surveying and wrangling dataset vaccine.csv

In [120]:
%%sql
DROP TABLE IF EXISTS vaccine;

CREATE TABLE vaccine (
Province_State text not null,
Date date not null,
Vaccine_Type text not null,
FIPS integer,
Country_Region text not null,
Lat numeric,
Long_ numeric,
Doses_alloc numeric,
Doses_shipped numeric,
Doses_admin numeric,
Stage_One_Doses numeric,
Stage_Two_Doses numeric,
Combined_Key text not null
)


 * postgresql://postgres@/final_project
Done.
Done.


[]

In [121]:
!pwd

/home/ubuntu


In [122]:
%%sql
COPY vaccine FROM '/home/ubuntu/vaccine.csv'
CSV
HEADER;
#here my location is /home/ubuntu, please change it accordingly as well.

 * postgresql://postgres@/final_project
115882 rows affected.
(psycopg2.errors.SyntaxError) syntax error at or near "#"
LINE 1: #here my location is /home/ubuntu, please change it accordin...
        ^

[SQL: #here my location is /home/ubuntu, please change it accordingly as well.]
(Background on this error at: http://sqlalche.me/e/f405)


#### check records of vaccine and surveying data

In [123]:
%%sql
SELECT COUNT(*) FROM vaccine;

 * postgresql://postgres@/final_project
1 rows affected.


[(115882,)]

In [124]:
%%sql
select * from vaccine limit 10;

 * postgresql://postgres@/final_project
10 rows affected.


[('Alabama', datetime.date(2020, 12, 10), 'All', 1, 'US', Decimal('32.3182'), Decimal('-86.9023'), None, None, None, None, None, 'Alabama, US'),
 ('Alabama', datetime.date(2020, 12, 10), 'Moderna', 1, 'US', Decimal('32.3182'), Decimal('-86.9023'), None, None, None, None, None, 'Alabama, US'),
 ('Alabama', datetime.date(2020, 12, 10), 'Pfizer', 1, 'US', Decimal('32.3182'), Decimal('-86.9023'), None, None, None, None, None, 'Alabama, US'),
 ('Alabama', datetime.date(2020, 12, 11), 'All', 1, 'US', Decimal('32.3182'), Decimal('-86.9023'), None, None, None, None, None, 'Alabama, US'),
 ('Alabama', datetime.date(2020, 12, 11), 'Moderna', 1, 'US', Decimal('32.3182'), Decimal('-86.9023'), None, None, None, None, None, 'Alabama, US'),
 ('Alabama', datetime.date(2020, 12, 11), 'Pfizer', 1, 'US', Decimal('32.3182'), Decimal('-86.9023'), None, None, None, None, None, 'Alabama, US'),
 ('Alabama', datetime.date(2020, 12, 12), 'All', 1, 'US', Decimal('32.3182'), Decimal('-86.9023'), None, None, None, None, None, 'Alabama, US'),
 ('Alabama', datetime.date(2020, 12, 12), 'Moderna', 1, 'US', Decimal('32.3182'), Decimal('-86.9023'), None, None, None, None, None, 'Alabama, US'),
 ('Alabama', datetime.date(2020, 12, 12), 'Pfizer', 1, 'US', Decimal('32.3182'), Decimal('-86.9023'), None, None, None, None, None, 'Alabama, US'),
 ('Alabama', datetime.date(2020, 12, 13), 'All', 1, 'US', Decimal('32.3182'), Decimal('-86.9023'), None, None, None, None, None, 'Alabama, US')]

#### check if vaccine has duplicated data

In [125]:
%%sql
select Province_State,Date,Vaccine_Type,FIPS,Country_Region,Lat,Long_,Doses_alloc,Doses_shipped,Doses_admin,Stage_One_Doses,Stage_Two_Doses,Combined_Key,
count(1) from vaccine
group by Province_State,Date,Vaccine_Type,FIPS,Country_Region,Lat,Long_,Doses_alloc,Doses_shipped,Doses_admin,Stage_One_Doses,Stage_Two_Doses,Combined_Key
having count(1)>1;

 * postgresql://postgres@/final_project
0 rows affected.


[]

#### check if vaccine has null values on attribute fips, lat and long_

In [126]:
%%sql
select count(*)
from vaccine
where fips is null or Lat is null or Long_ is null
;

 * postgresql://postgres@/final_project
1 rows affected.


[(10122,)]

In [127]:
%%sql
select *
from vaccine
where fips is null or Lat is null or Long_ is null
limit 5
;

 * postgresql://postgres@/final_project
5 rows affected.


[('Bureau of Prisons', datetime.date(2020, 12, 15), 'Pfizer', None, 'US', None, None, None, None, None, None, None, 'Bureau of Prisons, US'),
 ('Bureau of Prisons', datetime.date(2020, 12, 10), 'All', None, 'US', None, None, None, None, None, None, None, 'Bureau of Prisons, US'),
 ('Bureau of Prisons', datetime.date(2020, 12, 10), 'Moderna', None, 'US', None, None, None, None, None, None, None, 'Bureau of Prisons, US'),
 ('Bureau of Prisons', datetime.date(2020, 12, 10), 'Pfizer', None, 'US', None, None, None, None, None, None, None, 'Bureau of Prisons, US'),
 ('Bureau of Prisons', datetime.date(2020, 12, 11), 'All', None, 'US', None, None, None, None, None, None, None, 'Bureau of Prisons, US')]

##### from the above queries on the dataset vaccine, find fips, lat, long_ the three attributes should not be null but they have null values. all these data records are dirty data, they all need to update a default value or just delete them

## surveying dataset people_vaccinated.csv

In [128]:
!head people_vaccinated.csv | csvlook

| FIPS | Province_State | Country_Region |       Date |     Lat |    Long_ | Combined_Key | People_Fully_Vaccinated | People_Partially_Vaccinated |
| ---- | -------------- | -------------- | ---------- | ------- | -------- | ------------ | ----------------------- | --------------------------- |
| True | Alabama        | US             | 2020-12-10 | 32.318… | -86.902… | Alabama, US  |                         |                             |
| True | Alabama        | US             | 2020-12-11 | 32.318… | -86.902… | Alabama, US  |                         |                             |
| True | Alabama        | US             | 2020-12-12 | 32.318… | -86.902… | Alabama, US  |                         |                             |
| True | Alabama        | US             | 2020-12-13 | 32.318… | -86.902… | Alabama, US  |                         |                             |
| True | Alabama        | US             | 2020-12-14 | 32.318… | -86.902… | Alabama, US  |                     

In [129]:
import os

In [130]:
%time os.system("head -25000 people_vaccinated.csv | csvsort -c1 | head | csvlook")

CPU times: user 0 ns, sys: 4.12 ms, total: 4.12 ms
Wall time: 18.7 s


0

In [131]:
%time os.system("head -25000 people_vaccinated.csv | csvcut -c1 | csvsort -c1| head | csvlook")

CPU times: user 2.29 ms, sys: 2.26 ms, total: 4.55 ms
Wall time: 10.9 s


0

In [132]:
%%sql
DROP TABLE IF EXISTS people_vaccinated;

CREATE TABLE people_vaccinated (
FIPS integer,
Province_State text not null,
Country_Region text not null,
Date date not null,
Lat numeric,
Long_ numeric,
Combined_Key text not null,
People_Fully_Vaccinated numeric,
People_Partially_Vaccinated numeric
)

 * postgresql://postgres@/final_project
Done.
Done.


[]

In [133]:
!pwd

/home/ubuntu


In [134]:
%%sql
COPY people_vaccinated FROM '/home/ubuntu/people_vaccinated.csv'
CSV
HEADER;
#here my location is /home/ubuntu, please change it accordingly as well.

 * postgresql://postgres@/final_project
21372 rows affected.
(psycopg2.errors.SyntaxError) syntax error at or near "#"
LINE 1: #here my location is /home/ubuntu, please change it accordin...
        ^

[SQL: #here my location is /home/ubuntu, please change it accordingly as well.]
(Background on this error at: http://sqlalche.me/e/f405)


In [135]:
%%sql
select count(*) from people_vaccinated;

 * postgresql://postgres@/final_project
1 rows affected.


[(21372,)]

In [136]:
%%sql
select * from people_vaccinated limit 10;

 * postgresql://postgres@/final_project
10 rows affected.


[(1, 'Alabama', 'US', datetime.date(2020, 12, 10), Decimal('32.3182'), Decimal('-86.9023'), 'Alabama, US', None, None),
 (1, 'Alabama', 'US', datetime.date(2020, 12, 11), Decimal('32.3182'), Decimal('-86.9023'), 'Alabama, US', None, None),
 (1, 'Alabama', 'US', datetime.date(2020, 12, 12), Decimal('32.3182'), Decimal('-86.9023'), 'Alabama, US', None, None),
 (1, 'Alabama', 'US', datetime.date(2020, 12, 13), Decimal('32.3182'), Decimal('-86.9023'), 'Alabama, US', None, None),
 (1, 'Alabama', 'US', datetime.date(2020, 12, 14), Decimal('32.3182'), Decimal('-86.9023'), 'Alabama, US', None, None),
 (1, 'Alabama', 'US', datetime.date(2020, 12, 15), Decimal('32.3182'), Decimal('-86.9023'), 'Alabama, US', None, None),
 (1, 'Alabama', 'US', datetime.date(2020, 12, 16), Decimal('32.3182'), Decimal('-86.9023'), 'Alabama, US', None, None),
 (1, 'Alabama', 'US', datetime.date(2020, 12, 17), Decimal('32.3182'), Decimal('-86.9023'), 'Alabama, US', None, None),
 (1, 'Alabama', 'US', datetime.date(2020, 12, 18), Decimal('32.3182'), Decimal('-86.9023'), 'Alabama, US', None, None),
 (1, 'Alabama', 'US', datetime.date(2020, 12, 19), Decimal('32.3182'), Decimal('-86.9023'), 'Alabama, US', None, None)]

#### check if people_vaccinated has null values on attribute fips, lat and long_

In [137]:
%%sql
select count(*)
from people_vaccinated 
where fips is null or lat is null or long_ is null;

 * postgresql://postgres@/final_project
1 rows affected.


[(1716,)]

In [138]:
%%sql
select FIPS,Province_State,Country_Region,Date,Lat,Long_,Combined_Key,People_Fully_Vaccinated,People_Partially_Vaccinated, count(*)
from people_vaccinated 
group by FIPS,Province_State,Country_Region,Date,Lat,Long_,Combined_Key,People_Fully_Vaccinated,People_Partially_Vaccinated
having count(*)>1
order by FIPS,Province_State,Country_Region,Date,Lat,Long_,Combined_Key,People_Fully_Vaccinated,People_Partially_Vaccinated;

 * postgresql://postgres@/final_project
0 rows affected.


[]

#### deleting the fips is null, lat is null or long_ is null records

In [139]:
%%sql
delete from people_vaccinated where fips is null or lat is null or long_ is null;

 * postgresql://postgres@/final_project
1716 rows affected.


[]

### surveying dataset testing.csv

In [140]:
%%sql
DROP TABLE IF EXISTS testing;

CREATE TABLE testing (
date varchar(20) not null,
state text not null,
cases_conf_probable numeric,
cases_confirmed numeric,
cases_probable numeric,
tests_viral_positive numeric,
tests_viral_negative numeric,
tests_viral_total numeric,
tests_antigen_positive numeric,
tests_antigen_total numeric,
people_viral_positive numeric,
people_viral_total numeric,
people_antigen_positive numeric,
people_antigen_total numeric,
encounters_viral_total numeric,
tests_combined_total numeric
)

 * postgresql://postgres@/final_project
Done.
Done.


[]

In [141]:
!pwd

/home/ubuntu


In [142]:
%%sql
COPY testing FROM '/home/ubuntu/testing.csv'
CSV
HEADER;
#here my location is /home/ubuntu, please change it accordingly as well.

 * postgresql://postgres@/final_project
35508 rows affected.
(psycopg2.errors.SyntaxError) syntax error at or near "#"
LINE 1: #here my location is /home/ubuntu, please change it accordin...
        ^

[SQL: #here my location is /home/ubuntu, please change it accordingly as well.]
(Background on this error at: http://sqlalche.me/e/f405)


In [143]:
%%sql
select count(*) from testing;

 * postgresql://postgres@/final_project
1 rows affected.


[(35508,)]

In [144]:
%%sql
select * from testing limit 10;

 * postgresql://postgres@/final_project
10 rows affected.


[('3/6/2020', 'AK', None, None, None, None, None, Decimal('8'), None, None, None, None, None, None, None, Decimal('8')),
 ('3/7/2020', 'AK', None, None, None, None, None, Decimal('12'), None, None, None, None, None, None, None, Decimal('12')),
 ('3/8/2020', 'AK', None, None, None, None, None, Decimal('14'), None, None, None, None, None, None, None, Decimal('14')),
 ('3/9/2020', 'AK', None, None, None, None, None, Decimal('23'), None, None, None, None, None, None, None, Decimal('23')),
 ('3/10/2020', 'AK', None, None, None, None, None, Decimal('23'), None, None, None, None, None, None, None, Decimal('23')),
 ('3/11/2020', 'AK', None, None, None, None, None, Decimal('46'), None, None, None, None, None, None, None, Decimal('46')),
 ('3/12/2020', 'AK', None, None, None, None, None, Decimal('46'), None, None, None, None, None, None, None, Decimal('46')),
 ('3/13/2020', 'AK', None, None, None, None, None, Decimal('60'), None, None, None, None, None, None, None, Decimal('60')),
 ('3/14/2020', 'AK', None, None, None, None, None, Decimal('144'), None, None, None, None, None, None, None, Decimal('144')),
 ('3/15/2020', 'AK', None, None, None, None, None, Decimal('144'), None, None, None, None, None, None, None, Decimal('144'))]

#### alter table tesing: date attribute's type is date.

In [145]:
%%sql
update testing set date=to_char(to_date(date,'MM/DD/YYYY'),'YYYY-MM-DD');

 * postgresql://postgres@/final_project
35508 rows affected.


[]

In [146]:
%%sql
alter table testing alter column date type date using date ::date;

 * postgresql://postgres@/final_project
Done.


[]

In [147]:
%%sql
select * from testing limit 10;

 * postgresql://postgres@/final_project
10 rows affected.


[(datetime.date(2020, 3, 7), 'ME', Decimal('0'), None, None, None, None, None, None, None, Decimal('0'), None, None, None, None, None),
 (datetime.date(2020, 3, 1), 'MI', Decimal('14'), None, None, None, None, None, None, None, Decimal('14'), None, None, None, None, None),
 (datetime.date(2020, 3, 11), 'AK', None, None, None, None, None, Decimal('46'), None, None, None, None, None, None, None, Decimal('46')),
 (datetime.date(2020, 3, 12), 'AK', None, None, None, None, None, Decimal('46'), None, None, None, None, None, None, None, Decimal('46')),
 (datetime.date(2020, 3, 13), 'AK', None, None, None, None, None, Decimal('60'), None, None, None, None, None, None, None, Decimal('60')),
 (datetime.date(2020, 3, 14), 'AK', None, None, None, None, None, Decimal('144'), None, None, None, None, None, None, None, Decimal('144')),
 (datetime.date(2020, 3, 15), 'AK', None, None, None, None, None, Decimal('144'), None, None, None, None, None, None, None, Decimal('144')),
 (datetime.date(2020, 3, 16), 'AK', None, None, None, None, None, Decimal('144'), None, None, None, None, None, None, None, Decimal('144')),
 (datetime.date(2020, 5, 16), 'AS', Decimal('0'), None, None, None, None, Decimal('105'), None, None, Decimal('0'), None, None, None, None, Decimal('105')),
 (datetime.date(2020, 5, 17), 'AS', Decimal('0'), None, None, None, None, Decimal('105'), None, None, Decimal('0'), None, None, None, None, Decimal('105'))]

#### for the testing table is a statistical table so the null value all should update to 0

In [148]:
%%sql
update testing set cases_conf_probable=0 where cases_conf_probable is null;

 * postgresql://postgres@/final_project
287 rows affected.


[]

In [149]:
%%sql
update testing set cases_confirmed=0 where cases_confirmed is null;

 * postgresql://postgres@/final_project
22663 rows affected.


[]

In [150]:
%%sql
update testing set cases_probable=0 where cases_probable is null;

 * postgresql://postgres@/final_project
22664 rows affected.


[]

In [151]:
%%sql
update testing set tests_viral_positive=0 where tests_viral_positive is null;

 * postgresql://postgres@/final_project
19012 rows affected.


[]

In [152]:
%%sql
update testing set tests_viral_negative=0 where tests_viral_negative is null;

 * postgresql://postgres@/final_project
25683 rows affected.


[]

In [153]:
%%sql
update testing set tests_viral_total=0 where tests_viral_total is null;

 * postgresql://postgres@/final_project
8544 rows affected.


[]

In [154]:
%%sql
update testing set tests_antigen_positive=0 where tests_antigen_positive is null;

 * postgresql://postgres@/final_project
29465 rows affected.


[]

In [155]:
%%sql
update testing set tests_antigen_total=0 where tests_antigen_total is null;

 * postgresql://postgres@/final_project
25716 rows affected.


[]

In [156]:
%%sql
update testing set people_viral_positive=0 where people_viral_positive  is null;

 * postgresql://postgres@/final_project
3412 rows affected.


[]

In [157]:
%%sql
update testing set people_viral_total=0 where people_viral_total is null;

 * postgresql://postgres@/final_project
19810 rows affected.


[]

In [158]:
%%sql
update testing set people_antigen_positive =0 where people_antigen_positive is null;

 * postgresql://postgres@/final_project
34086 rows affected.


[]

In [159]:
%%sql
update testing set people_antigen_total=0 where people_antigen_total is null;

 * postgresql://postgres@/final_project
33423 rows affected.


[]

In [160]:
%%sql
update testing set encounters_viral_total=0 where encounters_viral_total is null;

 * postgresql://postgres@/final_project
25974 rows affected.


[]

In [161]:
%%sql
update testing set tests_combined_total=0 where tests_combined_total is null;

 * postgresql://postgres@/final_project
200 rows affected.


[]

## Ask 3 - Attestation (-10% for non-compliance)

All project team members must contribute meaningfully to the project and have complete knowledge on any part of it.

Each group member MUST attest to his/her individual contributions in coding, and presentation write up. 

Penalty: -10% for non-compliance

## Ask 4 - Analytical Questions (15%)

Communicate among your group members and produce analytical questions you want to answer with the data. Minimum of 4 major questions are required. 

You may consider me as your Sponsor. Arrange a 30 min meeting with me to discuss this part, if necessary. 

Expectation: A list of analytical questions that your model can answer. Check https://coronavirus.jhu.edu
- ”tracking” -> “Critical Trends” or U.S. Map 
- “Vaccines” -> “Tracker” for ideas 

## Ask 5 - Designing a Dimensional Database (25%)

Design a dimensional database to cover “Cases and Deaths”, “Vaccinations” and “Testing” datasets (Daily and for U.S.). I expect that your design will allow all other groups to answer most of their questions. 

Be precise and document your design correctly. 

Expectation: A star schema diagram detailing the design. Would be nice to produce an image of the design and added it to your code.

## Ask 6 - Writing Queries for Analytical Questions (35%) 

Write queries to answer your analytical questions.

Produce charts and make sure to include your code or a copy of chart (if you use outside Jupyter notebook tools). 

You are free to use any tool for visualization (e.g. tableau, ggplot, matplotlib,..).

Expectation: set of queries, results and charts. Those should be compatible with your presentation.